src/graph/sampler.cc

```diff
@@ -997,6 +998,124 @@ Subgraph NegEdgeSubgraph(int64_t num_tot_nodes, const Subgraph &pos_subg,
   return neg_subg;
 }
 
+Subgraph PBGNegEdgeSubgraph(int64_t num_tot_nodes, const Subgraph &pos_subg,
+                            const std::string &neg_mode,
+                            int neg_sample_size, bool is_multigraph,
+                            bool exclude_positive) {
+  std::vector<IdArray> adj = pos_subg.graph->GetAdj(false, "coo");
+  IdArray coo = adj[0];
+  int64_t num_pos_edges = coo->shape[0] / 2;
+
+  int64_t chunk_size = neg_sample_size;
+  // If num_pos_edges isn't divisible by chunk_size, the actual number of chunks
+  // is num_chunks + 1 and the last chunk size is last_chunk_size.
+  // Otherwise, the actual number of chunks is num_chunks, the last chunk size
+  // is 0.
+  int64_t num_chunks = num_pos_edges / chunk_size;
+  int64_t last_chunk_size = num_pos_edges - num_chunks * chunk_size;
+
+  // The number of negative edges.
+  int64_t num_neg_edges = neg_sample_size * chunk_size * num_chunks;
+  int64_t num_neg_edges_last_chunk = neg_sample_size * last_chunk_size;
+  int64_t num_all_neg_edges = num_neg_edges + num_neg_edges_last_chunk;
+
+  // We should include the last chunk.
+  if (last_chunk_size > 0)
+    num_chunks++;
+
+  IdArray neg_dst = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+  IdArray neg_src = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+  IdArray neg_eid = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+  IdArray induced_neg_eid = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+
+  // These are vids in the positive subgraph.
+  const dgl_id_t *dst_data = static_cast<const dgl_id_t *>(coo->data);
+  const dgl_id_t *src_data = static_cast<const dgl_id_t *>(coo->data) + num_pos_edges;
+  const dgl_id_t *induced_vid_data = static_cast<const dgl_id_t *>(pos_subg.induced_vertices->data);
+  const dgl_id_t *induced_eid_data = static_cast<const dgl_id_t *>(pos_subg.induced_edges->data);
+  size_t num_pos_nodes = pos_subg.graph->NumVertices();
+  std::vector<size_t> pos_nodes(induced_vid_data, induced_vid_data + num_pos_nodes);
+
+  dgl_id_t *neg_dst_data = static_cast<dgl_id_t *>(neg_dst->data);
+  dgl_id_t *neg_src_data = static_cast<dgl_id_t *>(neg_src->data);
+  dgl_id_t *neg_eid_data = static_cast<dgl_id_t *>(neg_eid->data);
+  dgl_id_t *induced_neg_eid_data = static_cast<dgl_id_t *>(induced_neg_eid->data);
+
+  bool neg_head = (neg_mode == "head");
```

Comment:
Can it be a global MACRO to handle sample mode?

Reply:
what do you mean by "global macro"?

Reply:
like:
```
bool neg_head = IS_PBG_HEAD_MODE(neg_mode)

#def IS_PBG_HEAD_MODE(neg_mode)  \
    (neg_mode == "head")
```

This can avoid possible inconsistency if we use bool neg_head = (neg_mode == "head"); in multiple code pieces.